src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,160 @@
+import React, { CSSProperties, useCallback, useEffect, useMemo } from 'react';
+import { CSSTransition } from 'react-transition-group';
+import classNames from 'classnames';
+import { ClassValue } from 'classnames/types';
+
+import ReactDOM from 'react-dom';
+import styles from './Modal.module.css';
+
+type Size = 'sm' | 'md' | 'lg';
+type Level = 'primary' | 'warning' | 'info' | 'danger';
+
+interface ModalProps {
+  children?: React.ReactNode;
+  className?: ClassValue;
+  isOpen?: boolean;
+  onClose?: () => void;
+  size?: Size;
+  style?: CSSProperties;
+}
+
+type ModalChildProps = Pick<ModalProps, 'children' | 'className'>;
+type ModalHeaderProps = Pick<ModalProps, 'children' | 'className'> & {
+  level?: Level;
+};
+
+type ModalType = React.FC<ModalProps> & {
+  Header: React.FC<ModalHeaderProps>;
+  Footer: React.FC<ModalChildProps>;
+  Body: React.FC<ModalChildProps>;
+};
+
+const Modal: ModalType = ({
+  children,
+  className: passedClassName = '',
+  isOpen = false,
+  onClose = () => {},
+  size = 'md',
+  style: passedStyle = {},
+  ...passedProps
+}) => {
+  const handleEscape = useCallback(
+    (e: KeyboardEvent) => {
+      if (isOpen && e.code === 'Escape' && onClose) {
+        onClose();
+      }
+    },
+    [isOpen, onClose]
+  );
+
+  useEffect(() => {
+    if (isOpen) {
+      window.addEventListener('keyup', handleEscape);
+    }
+
+    if (!isOpen) {
+      window.removeEventListener('keyup', handleEscape);
+    }
+
+    return () => {
+      if (isOpen) window.removeEventListener('keyup', handleEscape);
+    };
+  }, [handleEscape, isOpen]);
+
+  useEffect(() => {
+    if (isOpen) {
+      document.body.className = classNames(
+        document.body.className,
+        styles.bodyHasOpenModal
+      );
+    } else {
+      document.body.className = document.body.className
+        .replace(styles.bodyHasOpenModal, '')
+        .trim();
+    }
+  }, [isOpen]);
+
+  const modalContentStyle = useMemo(
+    () =>
+      classNames(
+        styles.modalContent,
+        {
+          [styles[`size--${size}`]]: size,
+        },
+        passedClassName
+      ),
+    [passedClassName, size]
+  );
+  return ReactDOM.createPortal(
+    <CSSTransition
+      in={isOpen}
+      unmountOnExit
+      timeout={{ enter: 0, exit: 300 }}
+      classNames={{
+        enterDone: styles.enterDone,
+        exit: styles.exit,
+      }}
+    >
+      <div
+        role="presentation"
+        {...passedProps}
+        style={passedStyle}
+        className={styles.modal}
+        onClick={onClose}
```

Comment:
Looking at an article about a11y in modals, I like the idea of creating a modal context and making the `onClose` function available to children that way.
https://medium.com/@seif_ghezala/how-to-create-an-accessible-react-modal-5b87e6a27503